### PR TITLE
add duplicate email feature

### PR DESCRIPTION
**Pull request by picsoung**
to solve #18 

Add a "duplicate" menu option, to duplicate email template.
With an intermediary screen to choose email name

### CODE REVIEW

File: server/storage.js [hunk 1]
```diff
@@ -82,4 +83,25 @@ app.post('/delete', wrap(async (req, res) => {
...
+app.post('/duplicate', wrap(async (req, res) => {
+  var content_html = await fs.readFile(path.join('./emails/', req.body.id+'.html'), {encoding: 'utf-8'});
+  var content_json = await fs.readFile(path.join('./emails/', req.body.id+'.json'), {encoding: 'utf-8'});
+
+  var metadata = JSON.parse(content_json).metadata;
+  metadata.name = req.body.email_name; //new name
+
+  var uuid = (req.body.uuid || newUuid());
+  await req.body.uuid || redis.lpushAsync(`mosaico:${req.user}:emails`, uuid);
+  await fs.writeFile(`./emails/${uuid}.html`, content_html);
```
Comment (guilro) [inline server/storage.js]: why not just pipe the first file into the new file ?

File: server/storage.js [hunk 2]
```diff
@@ -82,4 +83,25 @@ app.post('/delete', wrap(async (req, res) => {
...
+ */
+app.post('/duplicate', wrap(async (req, res) => {
+  var content_html = await fs.readFile(path.join('./emails/', req.body.id+'.html'), {encoding: 'utf-8'});
+  var content_json = await fs.readFile(path.join('./emails/', req.body.id+'.json'), {encoding: 'utf-8'});
+
+  var metadata = JSON.parse(content_json).metadata;
+  metadata.name = req.body.email_name; //new name
+
+  var uuid = (req.body.uuid || newUuid());
+  await req.body.uuid || redis.lpushAsync(`mosaico:${req.user}:emails`, uuid);
```
Comment (guilro) [inline server/storage.js]: Why keeping this condition ? We always have to generate a new uuid.

File: server/storage.js [hunk 3]
```diff
@@ -82,4 +83,24 @@ app.post('/delete', wrap(async (req, res) => {
...
 }));
 
+/**
+ * Duplicate the mail in redis db
+ */
+app.post('/duplicate', wrap(async (req, res) => {
+  var content_json = await fs.readFile(path.join('./emails/', req.body.id+'.json'), {encoding: 'utf-8'});
+
+  var metadata = JSON.parse(content_json).metadata;
+  metadata.name = req.body.email_name; //new name
```
Comment (guilro) [inline server/storage.js]: Where is the body posted ?

**Comment by picsoung** [comment]
Thanks @guilro for the comments :)
You are right, I corrected the lines you were mentioning.

The screen to select name appears when click on `duplicate` link.
duplicate view is in `server/views/duplicate.pug` I forgot to add it on my first commit 🙈

**Comment by guilro** [comment]
Thanks a lot ! @AlexisJC can you test ?

**Comment by AlexisJC** [comment]
Hi picsoung, when I want /view a duplicate mail I get "[object Promise]". But when i check with /edit is good, we have new duplicate mail from the source. So, I let you see how fix this, ty :)

File: server/storage.js [hunk 4]
```diff
@@ -82,4 +83,24 @@ app.post('/delete', wrap(async (req, res) => {
...
+
+  var uuid = newUuid();
+  await redis.lpushAsync(`mosaico:${req.user}:emails`, uuid);
+  await fs.writeFile(`./emails/${uuid}.html`, fs.readFile(path.join('./emails/', req.body.id+'.html'), {encoding: 'utf-8'}));
```
Comment (guilro) [inline server/storage.js]: I guess this line does not work. fs.readFile is a promise, not the content of the file. You should use .pipe()

Comment (AlexisJC) [inline server/storage.js]: var test = await fs.readFile(path.join(`./emails/${req.body.id}.html`));
await fs.writeFile(`./emails/${uuid}.html`, test , {encoding: 'utf-8'});

This work, but it's better to use : http://stackoverflow.com/questions/11293857/fastest-way-to-copy-file-in-node-js

Comment (guilro) [inline server/storage.js]: Who do the change ? @picsoung can you amend your PR ?

Comment (picsoung) [inline server/storage.js]: and voila :)
pushed changes.
Thanks for feedback. No more quick push while at work, I promise.

**Comment by guilro** [comment]
Squashed and merged manually ! Thanks :)

**Pull request closed without merging**